src/liballoc/arc.rs

```diff
@@ -1061,9 +1061,36 @@ impl<T: ?Sized> Weak<T> {
     }
 
     #[inline]
-    fn inner(&self) -> &ArcInner<T> {
+    fn inner(&self) -> Option<&ArcInner<T>> {
         // See comments above for why this is "safe"
-        unsafe { self.ptr.as_ref() }
+        //
+        // Note that the `Option` being returned here is pretty tricky, but is
+        // in relation to #48493. You can create an instance of `Weak<!>` which
+        // internally contains `NonNull<ArcInner<!>>`. Now the layout of
+        // `ArcInner<!>` directly embeds `!` so rustc correctly deduces that the
+        // size of `ArcInner<!>` is zero. This means that `Box<ArcInner<!>>`
+        // which we create in `Weak::new()` is actually the pointer 1 (pointers
+        // to ZST types are currently `1 as *mut _`).
+        //
+        // As a result we may not actually have an `&ArcInner<T>` to hand out
+        // here. That type can't actually exist for all `T`, such as `!`, so we
+        // can only hand out a safe reference if we've actually got one to hand
+        // out, aka if the size of the value behind the pointer is nonzero.
+        //
+        // Note that this adds an extra branch on methods like `clone` with
+        // trait objects like `Weak<Any>`, but we should be able to recover the
+        // original performance as soon as we can change `ArcInner` to store a
+        // `MaybeInitialized<!>` internally instead of a `!` directly. That way
+        // our allocation will *always* be allocated and we won't need this
+        // branch.
+        unsafe {
+            let ptr = self.ptr.as_ref();
+            if mem::size_of_val(ptr) == 0 {
```

Comment:
This is still inherently bogus since it's creating a reference to an uninhabited type, right?

Reply:
I think we're actually safe here in the sense that we're not telling LLVM this is illegal today and this compiles ok, but it's certainly standing on quite the precipice. If you'd prefer though I could probably add enough conditions here to skip this check if `T` isn't a dynamically sized type and only run this check if it's a DST.